internal/server/readhandler.go

```diff
@@ -13,66 +13,99 @@ import (
 
 	"github.com/edgexfoundry/device-opcua-go/pkg/result"
 	sdkModel "github.com/edgexfoundry/device-sdk-go/v3/pkg/models"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/clients/logger"
 	"github.com/gopcua/opcua"
 	"github.com/gopcua/opcua/ua"
 )
 
-func (s *Server) ProcessReadCommands(reqs []sdkModel.CommandRequest) ([]*sdkModel.CommandValue, error) {
-	var responses = make([]*sdkModel.CommandValue, len(reqs))
+type ResultToRequest map[int][]int
 
-	for i, req := range reqs {
-		res, err := s.handleReadCommandRequest(req)
-		if err != nil {
-			s.sdk.LoggingClient().Errorf("Driver.HandleReadCommands: Handle read commands failed: %v", err)
-			return responses, err
+func (rr ResultToRequest) buildCommandValues(reqs []sdkModel.CommandRequest, resp *ua.ReadResponse, logger logger.LoggingClient) []*sdkModel.CommandValue {
+	responses := make([]*sdkModel.CommandValue, len(reqs))
+	for i := 0; i < len(resp.Results); i++ {
+		if resp.Results[i].Status != ua.StatusOK {
+			logger.Debugf("Driver.handleReadCommands: Status not OK: %v", resp.Results[i].Status)
+			continue
+		}
+
+		variant := resp.Results[i].Value
+		if variant == nil || variant.Value() == nil {
+			continue
+		}
+
+		if reqIndexes, ok := rr[i]; ok {
+			for _, reqIndex := range reqIndexes {
+				var err error
+				if responses[reqIndex], err = result.NewResult(reqs[i], variant.Value()); err != nil {
+					logger.Errorf("Driver.handleReadCommands: Error: %v", err)
+				}
+			}
 		}
-		s.sdk.LoggingClient().Infof("Read command finished: %v", res)
-		responses[i] = res
 	}
 
-	return responses, nil
+	return responses
 }
 
-func (s *Server) handleReadCommandRequest(req sdkModel.CommandRequest) (*sdkModel.CommandValue, error) {
-	if _, isMethod := req.Attributes[METHOD]; isMethod {
-		return nil, fmt.Errorf("not allowed to call command on method: %s", req.DeviceResourceName)
+func buildNodesToReadRequest(reqs []sdkModel.CommandRequest) (nodesToRead []*ua.ReadValueID, resultToRequest ResultToRequest, err error) {
+	nodesToRead = make([]*ua.ReadValueID, 0, len(reqs))
+	resultToRequest = make(map[int][]int, len(reqs))
+	nodesIdToResultIndex := make(map[string]int, len(reqs))
+
+	for reqIndex, req := range reqs {
+		if _, isMethod := req.Attributes[METHOD]; isMethod {
+			err := fmt.Errorf("not allowed to call command on method: %s", req.DeviceResourceName)
+			return nil, nil, err
+		}
```

Comment:
Could you do this instead?
```suggestion
		if _, isMethod := req.Attributes[METHOD]; isMethod {
			return nil, nil, fmt.Errorf("not allowed to call command on method: %s", req.DeviceResourceName)
		}
```